### PR TITLE
Compatibility with Autofac 5

### PR DESCRIPTION
**Pull request by alistairjevans**
Needed some changes here because of the immutable registry.

I've removed the Provide methods that allows people to specify exact implementations/instances (which was modifying the registry).

If there are any exact implementations/instances, those can be provided in the configuration callback for the AutoMock as with any other custom implementations.  The documentation will need updating.

Basically, instead of:

```csharp
[Test]
public void Test()
{
  using (var mock = AutoMock.GetLoose())
  {
    var dependency = new Dependency();
    mock.Provide<IDependency>(dependency);

    // ...and the rest of the test.
  }
}
```
you would do this:
```csharp
[Test]
public void Test()
{
  var dependency = new Dependency();
  using (var mock = AutoMock.GetLoose(cfg => cfg.RegisterInstance(dependency).As<IDependency>())
  {
    // ...and the rest of the test.
  }
}
```

I've also added an extension method 'RegisterMock', to make it easier to register explicit mock objects in that configuration callback if needed.

### CODE REVIEW

**Comment by alistairjevans** [comment]
To whoever reviews this: I'm going to take a look at some of the other issues on this integration before we release 5.0 of this extension. See if we can't knock a couple of the issues on the head.

File: src/Autofac.Extras.Moq/AutoMock.cs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // This software is part of the Autofac IoC container
-// Copyright (c) 2013 Autofac Contributors
+// Copyright (c) 2020 Autofac Contributors
```
Comment (tillig) [inline src/Autofac.Extras.Moq/AutoMock.cs]: This should either stay 2013 (first year of publication) or be a range (2013-2020). [See explanation of copyright date here.](https://smallbusiness.chron.com/year-copyright-statement-mean-62027.html)

File: src/Autofac.Extras.Moq/MoqRegistrationHandler.cs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 // This software is part of the Autofac IoC container
-// Copyright (c) 2007 - 2008 Autofac Contributors
+// Copyright (c) 2020 Autofac Contributors
```
Comment (tillig) [inline src/Autofac.Extras.Moq/MoqRegistrationHandler.cs]: Another that should probably be 2007 or 2007-2020. (Did we really have this in 2007?)

File: src/Autofac.Extras.Moq/AutoMock.cs (hunk 1 above)
Comment (alistairjevans) [reply src/Autofac.Extras.Moq/AutoMock.cs]: Reverted

File: src/Autofac.Extras.Moq/MoqRegistrationHandler.cs (hunk 2 above)
Comment (alistairjevans) [reply src/Autofac.Extras.Moq/MoqRegistrationHandler.cs]: Reverted. Also, I doubt it...it wouldn't have been much use without AutoMock, which is apparently 6 years newer?

**Comment by alistairjevans** [comment]
Copyright changes reverted; apologies, force of habit from a work client that insists on getting copyright years updated.

**Comment by tillig** [comment]
I'll hold off on a release. Let me know when you're ready.

**Pull request merged**